// routeros-client/dist/RosApiMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RosApiMenu = void 0;
const RosApiCommands_1 = require("./RosApiCommands");
const RosApiModel_1 = require("./RosApiModel");
class RosApiMenu {
    /**
     * Creates a handler for the menus
     * and creates Models
     *
     * @param rosApi The raw api
     */
    constructor(rosApi) {
        /**
         * Toggles snake_case mode, defaults to camelCase
         */
        this.snakeCase = false;
        this.rosApi = rosApi;
    }
    /**
     * Returns a set of operations to do over
     * the menu provided
     *
     * @param path The menu you want to do actions
     */
    menu(path) {
        return new RosApiCommands_1.RosApiCommands(this.rosApi, path, this.snakeCase);
    }
    /**
     * Instead of transforming the routeros properties
     * to camelCase, transforms to snake_case
     */
    useSnakeCase() {
        this.snakeCase = true;
    }
    /**
     * Creates a model of pre-made actions to do
     * with an item from any menu that has lists
     *
     * @param item The item of a menu (Ex: a single firewall rule, or an IP address)
     */
    model(item) {
        if (Array.isArray(item)) {
            const items = [];
            for (const i of item) {
                items.push(new RosApiModel_1.RosApiModel(this.rosApi, i, this.snakeCase));
            }
            return items;
        }
        return new RosApiModel_1.RosApiModel(this.rosApi, item, this.snakeCase);
    }
}
exports.RosApiMenu = RosApiMenu;
