// routeros-client/dist/RosApiModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RosApiModel = void 0;
const RosApiCrud_1 = require("./RosApiCrud");
class RosApiModel extends RosApiCrud_1.RouterOSAPICrud {
    /**
     * Creates a model over the item provided,
     * giving options do edit, move, remove and etc
     *
     * @param rosApi the raw api
     * @param item the item to create a model
     * @param snakeCase if should turn properties to snake_case instead of camelCase
     */
    constructor(rosApi, item, snakeCase) {
        super(rosApi, item.$$path, snakeCase);
        this.originalItem = item;
        this.dissolveProperties();
    }
    /**
     * Alias to remove
     */
    delete() {
        return this.remove();
    }
    /**
     * Disable itself
     */
    disable() {
        return super.disable(this.originalItem.id).then((response) => {
            return this.refreshData(response);
        }).catch((err) => {
            return Promise.reject(err);
        });
    }
    /**
     * Enable itself
     */
    enable() {
        return super.enable(this.originalItem.id).then((response) => {
            return this.refreshData(response);
        }).catch((err) => {
            return Promise.reject(err);
        });
    }
    /**
     * Move itself above the id provided, or to the end if none provided
     *
     * @param to where to move to
     */
    move(to) {
        return super.moveEntry(this.originalItem.id, to).then((response) => {
            return this.refreshData(response);
        }).catch((err) => {
            return Promise.reject(err);
        });
    }
    /**
     * Unset properties of itself
     * @param properties properties to unset
     */
    unset(properties) {
        return super.unset(properties, this.originalItem.id).then((response) => {
            return this.refreshData(response);
        }).catch((err) => {
            return Promise.reject(err);
        });
    }
    /**
     * Updates itself with the data provided
     *
     * @param data new data to update to
     */
    update(data) {
        return super.update(data, this.originalItem.id).then((response) => {
            return this.refreshData(response);
        }).catch((err) => {
            return Promise.reject(err);
        });
    }
    /**
     * Removes itself
     */
    remove() {
        return super.remove(this.originalItem.id);
    }
    /**
     * Alias to update
     *
     * @param data new data to update to
     */
    set(data) {
        return this.update(data);
    }
    /**
     * Removes all dissolved item properties assigned to
     * this object
     */
    cleanDissolvedProperties() {
        for (const property in this.originalItem) {
            if (this.originalItem.hasOwnProperty(property)) {
                delete this[property];
            }
        }
    }
    /**
     * Dissolves all properties of the item into this
     * object
     */
    dissolveProperties() {
        for (const property in this.originalItem) {
            if (this.originalItem.hasOwnProperty(property)) {
                this[property] = this.originalItem[property];
            }
        }
    }
    /**
     * Refresh this object with new data of the item
     * after printing it again
     */
    refreshData(response) {
        this.cleanDissolvedProperties();
        this.originalItem = response;
        this.dissolveProperties();
        return Promise.resolve(this);
    }
}
exports.RosApiModel = RosApiModel;
