// routeros-client/dist/RouterOSClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouterOSClient = void 0;
const node_routeros_1 = require("node-routeros");
const RosApiMenu_1 = require("./RosApiMenu");
const events_1 = require("events");
class RouterOSClient extends events_1.EventEmitter {
    /**
     * Creates a client with the options provided,
     * so you are able to connect and input actions
     *
     * @param options Connection options
     */
    constructor(options) {
        super();
        this.options = options;
        this.rosApi = new node_routeros_1.RouterOSAPI(this.options);
    }
    /**
     * If it is connected or not
     */
    isConnected() {
        return this.rosApi.connected;
    }
    /**
     * Connects to the routerboard with the options provided
     */
    connect() {
        const api = this.api();
        if (this.rosApi.connected)
            return Promise.resolve(api);
        return this.rosApi.connect().then(() => {
            this.emit("connected", api);
            this.rosApi.once("error", (err) => this.emit("error", err));
            this.rosApi.once("close", () => this.emit("close"));
            return Promise.resolve(api);
        }).catch((err) => {
            this.emit("error", err);
            return Promise.reject(err);
        });
    }
    /**
     * Change current connection options
     * but it doesn't reconnect
     *
     * @param options Connection options
     */
    setOptions(options) {
        Object.assign(this.options, options);
        this.rosApi.setOptions(this.options);
        return this;
    }
    /**
     * Get an instance of the API to handle operations
     */
    api() {
        return new RosApiMenu_1.RosApiMenu(this.rosApi);
    }
    /**
     * Disconnect from the routerboard
     */
    disconnect() {
        return this.rosApi.close().then((api) => {
            this.emit("disconnected", this);
            return Promise.resolve(this);
        }).catch((err) => {
            return Promise.reject(err);
        });
    }
    /**
     * Alias to disconnect
     */
    close() {
        return this.disconnect();
    }
    /**
     * Alias to disconnect
     */
    end() {
        return this.disconnect();
    }
    /**
     * Return the options provided
     */
    getOptions() {
        return this.options;
    }
}
exports.RouterOSClient = RouterOSClient;
